File: ws-server/src/main/java/net/devopssolutions/demo/ws/server/component/RpcMethodDispatcher.java
package net.devopssolutions.demo.ws.server.component;

import lombok.Getter;
import lombok.extern.slf4j.Slf4j;
import net.devopssolutions.demo.ws.rpc.RpcMethod;
import net.devopssolutions.demo.ws.rpc.RpcMethodHandler;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.security.Principal;
import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;

@Slf4j
@Component
public class RpcMethodDispatcher {

    @Getter
    private Map<String, RpcMethodHandler> handlerMap;

    @Autowired
    private void init(List<RpcMethodHandler> handlers) {
        Function<RpcMethodHandler, String> keyMapper = rpcMethodHandler -> rpcMethodHandler.getClass().getAnnotation(RpcMethod.class).value();
        handlerMap = handlers.stream().collect(Collectors.toMap(keyMapper, rpcMethodHandler -> rpcMethodHandler));
    }

    @SuppressWarnings("unchecked")
    public void handle(String id, String method, Object params, Principal user) {
        RpcMethodHandler handler = handlerMap.get(method);
        if (handler == null) {
            log.warn("no handler for method: {} ", method);
            return;
        }

        handler.handle(id, params, user);
    }
}

File: ws-server/src/main/java/net/devopssolutions/demo/ws/server/component/RpcMethodDispatcher.kt
//package net.devopssolutions.demo.ws.server.component
//
//import mu.KLogging
//import net.devopssolutions.demo.ws.rpc.RpcMethod
//import net.devopssolutions.demo.ws.server.handler.RpcMethodHandler
//import org.springframework.stereotype.Component
//import reactor.core.publisher.Mono
//import java.security.Principal
//import kotlin.reflect.full.findAnnotation
//
//@Component
//open class RpcMethodDispatcher(handlers: List<RpcMethodHandler>) {
//    companion object : KLogging()
//
//    private val handlerMap: Map<String?, RpcMethodHandler> = handlers
//            .associateBy { it::class.findAnnotation<RpcMethod>()?.value?.method }
//
//    @SuppressWarnings("unchecked")
//    fun handle(sessionId: String, id: String, method: String, params: Any, user: Mono<Principal>) {
//        val rpcMethodHandler = handlerMap[method]
//        if (rpcMethodHandler != null) {
//            rpcMethodHandler.handle(sessionId, id, params, user)
//        } else {
//            logger.warn("no handler for method: {} ", method)
//        }
//    }
//}
